concat helper method to chain together multiple service implementations, #59

patriknw opened the pull request:
I have to test and add documentation, but wanted to see if you agree with the approach first.

Refs #59

raboof commented on runtime/src/main/scala/akka/grpc/javadsl/ServiceHandler.scala:
```diff
@@ -0,0 +1,54 @@
+import akka.http.javadsl.model.HttpRequest
+import akka.http.javadsl.model.HttpResponse
+import akka.http.javadsl.model.StatusCodes
+import akka.japi.{Function => JFunction}
```
Did you find a reason not to use `java.util.function.Function` here?

patriknw replied:
I'll change that occurrence also together with the other one

patriknw replied:
I'll do that in this PR

patriknw commented on codegen/src/main/twirl/templates/JavaServer/Handler.scala.txt:
```diff
@@ -24,18 +24,17 @@ import akka.grpc.javadsl.GrpcExceptionHandler;
-        result.completeExceptionally(new UnsupportedOperationException("Unexpected path"));
-        return result;
+      } else {
+        return notFound;
```
changed this to NOT_FOUND because I guess that's a more valid response if a single handler is used, and felt better to use that as the signal to continue with next handler in `concat` rather than using exception

patriknw commented on plugin-tester-scala/src/main/scala/example/myapp/helloworld/GreeterServer.scala:
```diff
@@ -38,8 +39,8 @@ object GreeterServer {
-    val service: HttpRequest => Future[HttpResponse] = GreeterServiceHandler(new GreeterServiceImpl(mat))
-      .orElse { case _  => Future.successful(HttpResponse(StatusCodes.NotFound)) }
+    val service: HttpRequest => Future[HttpResponse] =
+      ServiceHandler(GreeterServiceHandler(new GreeterServiceImpl(mat)))
```
would a single implementation be "wrong" if used without the NotFound?
```
val service: HttpRequest => Future[HttpResponse] = GreeterServiceHandler(new GreeterServiceImpl(mat))
```

It might be easy to miss that?
Should we change `GreeterServiceHandler.apply` to be a total function appending the NotFound by default, and make the partial function another method? `GreeterServiceHandler.partial`?

patriknw replied:
Then the `ServiceHandler.apply` that I added wouldn't be needed

ktoso commented on runtime/src/main/scala/akka/grpc/javadsl/ServiceHandler.scala:
```diff
@@ -0,0 +1,54 @@
+  private val notFound = Future.successful(HttpResponse.create().withStatus(StatusCodes.NOT_FOUND))
+
+  /**
+   * Creates a `HttpRequest` to `HttpResponse` handler that can be used in for example `Http.get(system).bindAndHandleAsync`
```
maybe as [[]] links? (the response/request)

patriknw commented:
> I think the utility looks useful.

I think that for javadsl it's essential, "impossible" to compose several otherwise

patriknw commented:
> As for naming, though we started with it perhaps 'concat' (as we've called it before) isn't great since unlike Route.concat it also ends with returning a 404 (which I think is useful).

`concatOrNotFound`?

patriknw commented on codegen/src/main/twirl/templates/JavaServer/Handler.scala.txt:
```diff
@@ -24,18 +24,17 @@ import akka.grpc.javadsl.GrpcExceptionHandler;
 
 public class @{service.name}HandlerFactory {
+  private static CompletionStage<HttpResponse> notFound = CompletableFuture.completedFuture(
+    HttpResponse.create().withStatus(StatusCodes.NOT_FOUND));
```
yes, this is for the path segments, the method match is still using UnsupportedOperationException

ktoso commented on runtime/src/main/scala/akka/grpc/javadsl/ServiceHandler.scala:
```diff
@@ -0,0 +1,54 @@
+    new JFunction[HttpRequest, CompletionStage[HttpResponse]] {
+      override def apply(req: HttpRequest): CompletionStage[HttpResponse] =
+        cont(req, notFound, handlers.toList).toJava
+    }
```
I see, since we generate functions and not `javadsl.Route`, looks good

ktoso commented:
> concatOrNotFound?

Sounds good to me

patriknw commented on codegen/src/main/twirl/templates/JavaServer/Handler.scala.txt:
```diff
@@ -24,18 +24,17 @@ import akka.grpc.javadsl.GrpcExceptionHandler;
 import static @{service.packageName}.@{service.name}.Serializers.*;
 
 public class @{service.name}HandlerFactory {
+  private static CompletionStage<HttpResponse> notFound = CompletableFuture.completedFuture(
```
thx

patriknw commented:
Updated with the things we discussed so far. Will continue with test and docs

patriknw commented:
Added documentation and test.
Also changed the implementation for javadsl to use plain CompletionStage.

This should be ready for final review now.

patriknw commented:
and the reason for the japi.Function was that bindAndHandleAsync expected that, so I reverted that change

raboof commented on runtime/src/main/scala/akka/grpc/javadsl/ServiceHandler.scala:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
+ */
```
I think sbt-headers expects a newline after this

raboof commented on sbt-plugin/src/sbt-test/gen-scala-server/02-multiple-services/src/main/scala/example/myapp/Main.scala:
```diff
@@ -18,13 +19,12 @@ object Main extends App {
-  val greeterHandler = GreeterServiceHandler(new GreeterServiceImpl)
+  val echoHandler = EchoServiceHandler.partial(new EchoServiceImpl)
+  val greeterHandler = GreeterServiceHandler.partial(new GreeterServiceImpl)
+  val serviceHandler = ServiceHandler.concatOrNotFound(echoHandler, greeterHandler)
```
:+1:

raboof commented on plugin-tester-java/src/main/java/example/myapp/CombinedServer.java:
```diff
@@ -0,0 +1,112 @@
+      });
+  }
+
+  private static HttpsConnectionContext serverHttpContext() throws Exception {
```
Yes, finding a nicer alternative for that is part of #89

patriknw commented:
Fixed headerCheck, rebased and squashed
